File: Java/src/Solution601_700/Solution697.java
package Solution601_700;

import java.util.HashMap;
import java.util.Map;

/**
 * @author windows
 */
public class Solution697 {
    public int findShortestSubArray(int[] nums) {
        HashMap<Integer, int[]> map = new HashMap<>();
        int maxCount = 0;
        for (int i = 0; i < nums.length; i++) {
            int[] ints = map.getOrDefault(nums[i], new int[3]);
            if (map.containsKey(nums[i])) {
                ints[2] = i;
            } else {
                ints[1] = i;
                ints[2] = i;
            }
            ints[0]++;
            map.put(nums[i], ints);
            maxCount = Math.max(maxCount, ints[0]);
        }
        int minLength = nums.length;
        for (Map.Entry<Integer, int[]> entry : map.entrySet()) {
            int[] value = entry.getValue();
            if (value[0] == maxCount){
                minLength = Math.min(value[2] - value[1] + 1, minLength);
            }
        }
        return minLength;
    }


    public static void main(String[] args) {
        Solution697 solution697 = new Solution697();
        System.out.println(solution697.findShortestSubArray(new int[]{2,1}));
    }
}

File: C++/src/Solution601_700/Solution697.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution697.h"
#include <unordered_map>

int Solution697::findShortestSubArray(vector<int> &nums) {
    unordered_map<int, vector<int>> hashtable;
    for (int i = 0; i < nums.size(); ++i) {
        if (!hashtable.count(nums[i])){
            hashtable[nums[i]] = {1, i, i};
        } else{
            hashtable[nums[i]][0]++;
            hashtable[nums[i]][2] = i;
        }
    }
    int max_value = 0, min_len = 0;
    for(auto& [_, value] : hashtable){
        if (value[0] > max_value){
            max_value = value[0];
            min_len = value[2] - value[1] + 1;
        }else if (value[0] == max_value){
            int temp = value[2] - value[1] + 1;
            min_len = min(temp, min_len);
        }
    }
    return min_len;
}
